nextcord/http.py

```diff
@@ -122,25 +138,331 @@ def bucket(self) -> str:
         return f"{self.channel_id}:{self.guild_id}:{self.path}"
 
 
-class MaybeUnlock:
-    def __init__(self, lock: asyncio.Lock) -> None:
-        self.lock: asyncio.Lock = lock
-        self._unlock: bool = True
+class RateLimitMigrating(DiscordException):
+    ...
+
+
+class IncorrectBucket(DiscordException):
+    ...
+
+
+class RateLimit:
+    """Used to time gate a large batch of requests to only allow X every Y seconds. Used via an async context manager.
+
+    NOT THREAD SAFE.
+
+    Parameters
+    ----------
+    time_offset: :class:`float`
+        Number in seconds to increase all timers by. Used for lag compensation.
+    """
+
+    _reset_ignore_threshold = 0.02  # Arbitrary number, feel free to tweak.
+    """Minimum reset - current time diff threshold. Any diff lower will be ignored.
+    Only used when use_reset_timestamp is enabled.
+    """
+
+    def __init__(self, time_offset: float = 0.0, *, use_reset_timestamp: bool = False) -> None:
+        self.limit: int = 1
+        """Maximum amount of requests before requests have to wait for the rate limit to reset."""
+        self.remaining: int = 1
+        """Remaining amount of requests before requests have to wait for the rate limit to reset."""
+        self.reset: datetime | None = None
```

Comment:
See earlier comments
```suggestion
        self.reset: Optional[datetime] = None
```